File: tests/pendings/monomorphization_bug/monomorphization_bug.c
#include "milone.h"

struct StringStringTuple2_;

void const* dup_(void const* arg_);

void const* bug_(void const* b_);

int milone_main(void);

struct StringStringTuple2_ {
    struct String t0;
    struct String t1;
};

void const* dup_(void const* arg_) {
    struct String x_ = (*(((struct String const*)arg_)));
    struct StringStringTuple2_ StringStringTuple2_ = (struct StringStringTuple2_){.t0 = x_, .t1 = x_};
    void const* box_ = milone_mem_alloc(1, sizeof(struct StringStringTuple2_));
    (*(((struct StringStringTuple2_*)box_))) = StringStringTuple2_;
    return box_;
}

void const* bug_(void const* b_) {
    void const* call_ = dup_(b_);
    return call_;
}

int milone_main(void) {
    printf("x1 = \'a\'\n");
    printf("x2 = \'a\'\n");
    return 0;
}

File: tests/pendings/monomorphization_bug/monomorphization_bug.fs
module rec monomorphization_bug.Program

type private Box<'T> = Box of 'T

let private dup (Box x) : Box<_ * _> = Box(x, x)

let private bug (b: Box<string>) : Box<string * string> =
  // Here `dup: Box<_> -> Box<_>`.
  // AutoBoxing pass modifies to `dup: obj -> obj`.
  // Monomorphization pass fails to restore type arguments of `dup<_>`.
  // (`dup<obj>` is unfortunately instantiated and runs into segmentation fault.)
  dup b

let main _ =
  // let (Box (x1, x2)) = bug (Box "a")

  // printfn "x1 = '%s'" x1
  // printfn "x2 = '%s'" x2

  // stub:
  printfn "x1 = 'a'"
  printfn "x2 = 'a'"
  0
